napari/_qt/menus/plugins_menu.py

```diff
@@ -51,48 +51,71 @@ def _remove_unregistered_widget(self, event):
                 self.removeAction(action)
                 self._win._remove_dock_widget(event=event)
 
-    def _add_registered_widget(self, event, call_all=False):
-        from ...plugins import menu_item_template, plugin_manager
+    def _add_registered_widget(self, event=None, call_all=False):
+        from itertools import chain
 
-        for hook_type, (plugin_name, widgets) in plugin_manager.iter_widgets():
-            if call_all or event.value == plugin_name:
-                multiprovider = len(widgets) > 1
-                if multiprovider:
-                    menu = NapariMenu(plugin_name, self)
-                    self.addMenu(menu)
-                else:
-                    menu = self
+        from ...plugins import plugin_manager
 
-                for wdg_name in widgets:
-                    key = (plugin_name, wdg_name)
-                    if multiprovider:
-                        action = QAction(wdg_name, parent=self)
+        _iterable: Iterator[Tuple[str, Tuple[str, Sequence[str]]]]
+        try:
+            import npe2
+        except ImportError:
+            _iterable = iter([])
+        else:
+            pm = npe2.PluginManager.instance()
+            wdgs: DefaultDict[str, List[str]] = DefaultDict(list)
+            for wdg_contrib in pm.iter_widgets():
+                wdgs[wdg_contrib.plugin_name].append(wdg_contrib.name)
+            _iterable = (('dock', x) for x in wdgs.items())
```

Comment:
this is the only new section here... the rest is just refactored to reduce code complexity